src/array_api_extra/_lib/_lazy.py

```diff
@@ -0,0 +1,403 @@
+        Any array objects in them will be converted to numpy when ``as_numpy=True``.
+
+    Returns
+    -------
+    Array | tuple[Array, ...]
+        The result(s) of `func` applied to the input arrays, wrapped in the same
+        array namespace as the inputs.
+        If shape is omitted or a single `tuple[int | None, ...]`, return a single array.
+        Otherwise, return a tuple of arrays.
+
+    Notes
+    -----
+    JAX
+        This allows applying eager functions to jitted JAX arrays, which are lazy.
+        The function won't be applied until the JAX array is materialized.
+        When running inside ``jax.jit``, `shape` must be fully known, i.e. it cannot
+        contain any `None` elements.
+
+        .. warning::
+
+            `func` must never raise inside ``jax.jit``, as the resulting behavior is
+            undefined.
+
+        Using this with `as_numpy=False` is particularly useful to apply non-jittable
+        JAX functions to arrays on GPU devices.
+        If ``as_numpy=True``, the :doc:`jax:transfer_guard` may prevent arrays on a GPU
+        device from being transferred back to CPU. This is treated as an implicit
+        transfer.
+
+    PyTorch, CuPy
+        If ``as_numpy=True``, these backends raise by default if you attempt to convert
+        arrays on a GPU device to NumPy.
+
+    Sparse
+        If ``as_numpy=True``, by default sparse prevents implicit densification through
+        :func:`numpy.asarray`. `This safety mechanism can be disabled
+        <https://sparse.pydata.org/en/stable/operations.html#package-configuration>`_.
+
+    Dask
+        This allows applying eager functions to dask arrays.
+        The dask graph won't be computed.
+
+        `lazy_apply` doesn't know if `func` reduces along any axes; also, shape
+        changes are non-trivial in chunked Dask arrays. For these reasons, all inputs
+        will be rechunked into a single chunk.
+
+        .. warning::
+
+           The whole operation needs to fit in memory all at once on a single worker.
+
+        The outputs will also be returned as a single chunk and you should consider
+        rechunking them into smaller chunks afterwards.
+
+        If you want to distribute the calculation across multiple workers, you
+        should use :func:`dask.array.map_blocks`, :func:`dask.array.map_overlap`,
+        :func:`dask.array.blockwise`, or a native Dask wrapper instead of
+        `lazy_apply`.
+
+    Dask wrapping around other backends
+        If ``as_numpy=False``, `func` will receive in input eager arrays of the meta
+        namespace, as defined by the ``._meta`` attribute of the input Dask arrays.
+        The outputs of `func` will be wrapped by the meta namespace, and then wrapped
+        again by Dask.
+
+    Raises
+    ------
+    jax.errors.TracerArrayConversionError
+        When ``xp=jax.numpy``, `shape` is unknown (it contains None on one or more axes)
```

Comment:
I had to think about this for a while, and I suspect other readers may have to do the same. The question that "shape is unknown" brings up is how that happens. I guess it's a combo of data-dependent-shape in the input array (e.g., it's a return value from `unique_values`) and a lack of input validation on `shape`?

Did you consider checking this (`None in shape and is_iax_namespace(xp)`) and raising a more informative error than `TracerArrayConversionError` within a`lazy_apply`? I'm asking because I remember the `jax` exception not being very informative.

Reply:
> I guess it's a combo of data-dependent-shape in the input array (e.g., it's a return value from unique_values) and a lack of input validation on shape?

No, JAX arrays can't have None in the shape today. Functions like `unique_values` without the bespoke `size=` parameter, as well as `a[mask]`, flat out crash inside jax.jit.
This is about *this* function creating an unknown shape. I'll update the docstring to clarify.

Trivial example that reimplements `jnp.unique_values`:

```python
y = xpx.lazy_apply(np.unique_values, x, shape=(None, ), as_numpy=True)
```

> Did you consider checking this (None in shape and is_iax_namespace(xp)) and raising a more informative error than TracerArrayConversionError within alazy_apply?

Good point. Fixed.

Reply:
Ah got it, thanks. Fix LGTM, so resolving this comment.